hack/make/test-unit: support coverprofile

AkihiroSuda opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**\- What I did**
Support coverprofiles for `hack/make/test-unit`.

**\- How I did it**
`test-integration-cli` uses `go_test_dir` for fetching the coverprofile.
Now test-unit utilizes `go_test_dir` as well.

**\- How to verify it**

```
$ make test-unit
$ ls bundles/latest/test-unit/coverprofiles/
docker-api
docker-api-client
...
```

**\- Description for the changelog**

Update #22899

TODOs suggested in #22899 (should be done in separate PRs after merging this PR)
- Update https://github.com/docker/opensource/blob/master/docs/project/test-and-docs.md
- Integrate  https://codecov.io
- Having Jenkins output the code coverage to its workspace

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp

AkihiroSuda commented:
The test failures seems related to my change. I'll look into this.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "coverprofile" git@github.com:AkihiroSuda/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented:
The `TestJSONFormatProgress` should be fixed now in https://github.com/docker/docker/pull/23113, so restarting CI

thaJeztah commented:
boo. looks like CI doesn't want to restart, @AkihiroSuda can you try rebasing and pushing again?

AkihiroSuda commented:
@thaJeztah OK, but I need to wait for #23120, #23152, and LLVM apt issue

AkihiroSuda commented:
@thaJeztah I think the PR is ready for review. The failure in win2lin seems not related to my change.

Note that the build time seems getting longer :disappointed: 
e.g. experimental took 1h35min https://jenkins.dockerproject.org/job/Docker-PRs-experimental/19367/

Maybe we can solve this performance regression by disabling `go test -c` and just doing `go test $someflags` at once for multiple packages.

thaJeztah commented:
Thanks! I removed the "failing-ci" label and restarted win2lin.

Yes, the tests are really getting long, perhaps we should consider something like that to see if it helps speeding up

thaJeztah commented:
ping @vdemeester @justincormack could you have a look at this one?

AkihiroSuda commented:
Amended the PR so as to compile test binaries at once for multiple packages.
See if it works..

thaJeztah commented:
Hm, something's failing;

```
11:42:00 ---> Making bundle: .ensure-nnp-test (in bundles/1.12.0-dev/test-integration-cli)
11:42:12 can't load package: package github.com/docker/docker: no buildable Go source files in /go/src/github.com/docker/docker
```

AkihiroSuda committed:
hack/make/test-unit: support coverprofile

e.g. for github.com/docker/api, the coverprofile is located on bundles/latest/test-unit/coverprofiles/docker-api

Update #22899

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

AkihiroSuda commented:
CI is passing, but I found that no coverprofile for integration-test is generated.. perhaps related to the handling of `-args` for `go test`...

Sorry for failing repeatedly..
Closing the PR temporarily.
I will reopen when the code is ready

thaJeztah commented:
@AkihiroSuda no worries, thanks in advance! Looking forward to an updated PR
